src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.ContextAwareSupport;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.NanoService;
+import foam.nanos.session.Session;
+
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.AND;
+
+public class AgentUserAuthService
+  extends    ContextAwareSupport
+  implements AgentAuthService, NanoService
+{
+  protected DAO userDAO_;
+  protected DAO groupDAO_;
+  protected DAO sessionDAO_;
+  protected DAO agentJunctionDAO_;
+
+  public AgentUserAuthService(X x) {
+    setX(x);
+  }
+
+  @Override
+  public void start() {
+    userDAO_     = (DAO) getX().get("localUserDAO");
+    groupDAO_    = (DAO) getX().get("groupDAO");
+    sessionDAO_  = (DAO) getX().get("sessionDAO");
+    agentJunctionDAO_  = (DAO) getX().get("agentJunctionDAO");
+  }
+
+  /* 
+    Sets the currently logged in user as an "agent" within the context &
+    sets "user" in the context to the passed in user. This allows users to
+    act on behalf of others while retaining information on the user.
+  */
+  public User ActAs(X x, User sudoUser) throws AuthenticationException {
+    User agent = (User) x.get("user");
+    User user = (User) userDAO_.find(sudoUser.getId());
+
+    if ( agent == null ) {
+      throw new AuthenticationException();
+    }
+
+    if ( user == null || ! user.getEnabled() ) {
+      throw new AuthenticationException("Super user doesn't exist.");
+    }
+
+    Group group = (Group) groupDAO_.find(user.getGroup());
+
+    if ( group == null ) {
+      throw new AuthenticationException("User must exist within a group.");
+    }
+
+    // Finds the AgentUserJunction object to see if user can log in as the passed in user.
+    // Source users are permissioned to act as target users, not visa versa.
+
+    UserUserJunction permissionJunction = (UserUserJunction) agentJunctionDAO_.find(AND(
+      EQ(UserUserJunction.SOURCE_ID, agent.getId()),
+      EQ(UserUserJunction.TARGET_ID, user.getId())
+    ));
+
+    if ( permissionJunction == null ) {
+      throw new AuthenticationException("You don't have access to act as the requested user.");
+    }
+
+    // Junction object contains a group which has a unique set of permissions specific to the relationship.
+    Group ActingWithinGroup = (Group) groupDAO_.find(permissionJunction.getGroup());
+
+    if ( group != null && ! group.getEnabled() ) {
+      throw new AuthenticationException("User group disabled");
+    }
+    
+    Session session = x.get(Session.class);
+    session.setUserId(user.getId());
+    session.setContext(session.getContext().put("user", user));
+    session.setContext(session.getContext().put("agent", agent));
+    session.setContext(session.getContext().put("group", ActingWithinGroup));
```

Comment:
I thought the point was to set the group on the user in the context to the group referenced by the junction, not put the group into the context directly. Everywhere in the code that we check a user's group we're first getting the user from the context and then looking up their group. All of those places would have to be updated to check the context directly instead if we were to do this.

That's my biggest question about this PR.

Reply:
The problem with that approach is if a user from the context is updated. The group appended to the context user would be updated to the user journal. This would cause users to be apart of groups they shouldn't be apart of. Regardless this implementation requires some sort of permitted update on super users in the future.

Also, there aren't many places that check groups other than the auth service and hopefully I've accounted for all the places where we have to set the group to the context.

Reply:
> The problem with that approach is if a user from the context is updated.

Right. One way we could solve this is to add a new decorator  to `userDAO` (or just modify AuthenticatedUserDAO) to prevent any updates to the current user if there's an agent in the context.

I think this is a better solution than changing the way we access a user's group everywhere in the system.

Example code:

```Java
User user = (User) x.get("user");
User agent = (User) x.get("agent");
User existingUser = (User) super.find_(x, obj);
boolean isUpdate = existingUser != null;
boolean isUpdatingSelf = isUpdate && existingUser.getId() == user.getId();
if ( isUpdatingSelf && agent != null ) {
  throw new AuthorizationException("You cannot update a user you are sudoed into.");
}
```